src/shogun/machine/gp/SingleFITCLaplacianBase.h

```diff
@@ -0,0 +1,269 @@
+	 */
+	virtual SGVector<float64_t> get_derivative_related_cov_diagonal();
+
+	/** helper function to compute matrices which are required to compute negative log marginal
+	 * likelihood derivatives wrt mean
+	 *
+	 * @param the explicit gradient wrt hyperparameter related to mean
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual float64_t get_derivative_related_mean(SGVector<float64_t> dmu);
+
+	/** helper function to compute matrices which are required to compute negative log marginal
+	 * likelihood derivatives wrt inducing features 
+	 *
+	 * Note that the kernel must support to compute the derivatives wrt inducing features
+	 *
+	 * @param the explicit gradient wrt hyperparameter related to inducing features
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_related_inducing_features(SGVector<float64_t> v,
+	SGMatrix<float64_t> BdK, const TParameter* param);
+
+	/** update alpha vector */
+	virtual void update_alpha()=0;
+
+	/** update cholesky matrix */
+	virtual void update_chol()=0;
+
+	/** update matrices which are required to compute negative log marginal
+	 * likelihood derivatives wrt hyperparameter
+	 */
+	virtual void update_deriv()=0;
+
+	/** returns derivative of negative log marginal likelihood wrt parameter of
+	 * CInferenceMethod class
+	 *
+	 * @param param parameter of CInferenceMethod class
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_inference_method(
+			const TParameter* param);
+
+	/** returns derivative of negative log marginal likelihood wrt parameter of
+	 * likelihood model
+	 *
+	 * @param param parameter of given likelihood model
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_likelihood_model(
+			const TParameter* param)=0;
+
+	/** returns derivative of negative log marginal likelihood wrt kernel's
+	 * parameter
+	 *
+	 * @param param parameter of given kernel
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_kernel(
+			const TParameter* param);
+
+	/** returns derivative of negative log marginal likelihood wrt mean
+	 * function's parameter
+	 *
+	 * @param param parameter of given mean function
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_mean(
+			const TParameter* param);
+
+	/** returns derivative of negative log marginal likelihood wrt inducing features (input)
+	 * Note that in order to call this method, kernel must support FITC inference,
+	 * which means derivatives wrt inducing features can be computed
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_inducing_features(const TParameter* param);
+
+	/* Note that alpha is NOT post.alpha */
```

Comment:
With all those vectors around, it might get a bit confusing what is what.

I suggest you write down the math of the thing somewhere, with well defined expressions, and then you can reference this. I.e. "this method computes the $L_R$ matrix of expression XYZ"

What do you think?